rt/features/metrics/src/main/java/org/apache/cxf/metrics/micrometer/MicrometerMetricsContext.java

```diff
@@ -0,0 +1,138 @@
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cxf.metrics.micrometer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.google.common.collect.Iterables;
+
+import org.apache.cxf.message.Exchange;
+import org.apache.cxf.message.Message;
+import org.apache.cxf.metrics.MetricsContext;
+import org.apache.cxf.metrics.micrometer.provider.TagsCustomizer;
+import org.apache.cxf.metrics.micrometer.provider.TagsProvider;
+import org.apache.cxf.metrics.micrometer.provider.TimedAnnotationProvider;
+
+import io.micrometer.core.annotation.Timed;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+
+public class MicrometerMetricsContext implements MetricsContext {
+
+    private final MeterRegistry registry;
+    private final TagsProvider tagsProvider;
+    private final TimedAnnotationProvider timedAnnotationProvider;
+    private final List<TagsCustomizer> tagsCustomizers;
+
+    private final String metricName;
+    private final boolean autoTimeRequests;
+
+    public MicrometerMetricsContext(MeterRegistry registry, TagsProvider tagsProvider,
+                                    TimedAnnotationProvider timedAnnotationProvider,
+                                    List<TagsCustomizer> tagsCustomizers, String metricName, boolean autoTimeRequests) {
+        this.registry = registry;
+        this.tagsProvider = tagsProvider;
+        this.timedAnnotationProvider = timedAnnotationProvider;
+        this.tagsCustomizers = tagsCustomizers;
+        this.metricName = metricName;
+        this.autoTimeRequests = autoTimeRequests;
+    }
+
+    @Override
+    public void start(Exchange ex) {
+        Message request = ex.getInMessage();
+        TimingContext timingContext = TimingContext.get(request);
+        if (timingContext == null) {
+            startAndAttachTimingContext(request);
+        }
+    }
+
+    @Override
+    public void stop(long timeInNS, long inSize, long outSize, Exchange ex) {
+        Message request = ex.getInMessage();
+        TimingContext timingContext = TimingContext.get(request);
+        if (timingContext == null) {
```

Comment:
The flow with nullable `timingContext` concerns a bit. It should not happen, but assuming it does, we are going to record effectively zeroed samples (since start and stop time would be mostly equal), which basically would screw / mess the metric. I would suggest to skip the `record` call in this case, or if we could, provide our own sample (since we know the time it took, `timeInNS`), what do you think?

Reply:
Good catch!
Yes, it should not happen.
I've checked `Timer#Sample` class. On the one hand we can use `timeInNS` to calculate "original" timestamp from current time and this one. On the other hand, we have to create a custom implementation of `Clock#monotonicTime` for creating a "start" `Timer#Sample`. As I see, this is used internally and it has a mock implementation for testing. 
In my opinion we should skip the record call in this case, and log this case as a warning.

Reply:
👍 agree

Reply:
Yes, sure.

Reply:
Done.